File: web/src/lib/families/resealTrigger.ts
// FSH-02's lazy-reseal TRIGGER (Plan 30-13) -- the fallback half of
// 30-DECISION-FSH-02.md's hybrid mechanism, and the thing that actually
// delivers a family-wide key to a newcomer whose invite was generated BEFORE
// the share existed (the invite-time wrap of 30-07/30-08 covers the other
// half, and cannot cover this one: an `invitations` row's payload is fixed at
// INSERT time and never re-computed).
//
// Cadence mirrors `identity/ensure.ts`'s `publishOnUnlock` shape: it runs
// opportunistically off the existing unlock/sync cycle, is never awaited by
// anything on the critical path, and tolerates its own failure completely.
//
// THE TRIGGER SET DELIBERATELY INCLUDES THE SHARER. 30-DECISION-FSH-02.md's
// one refinement over the starting hypothesis is that the trigger fires on
// "any current keyholder's own session" -- with NO special-casing that
// excludes the member who created the share. The sharer always already holds
// a usable key for their own family-wide share by construction (the existing
// multi-recipient fan-out grants every current member one at creation time),
// so including them shrinks the "nobody else ever opens the app" failure case
// for free. There is deliberately no `recipient_user_id !== me` style guard
// anywhere in this file.
//
// One query, two consumers: this module reads
// `familyWidePending.ts`'s synchronous `getFamilyWidePendingSnapshot()` and
// NEVER calls `getFamilyWidePending()` itself -- `sync.ts::pullOnce` performs
// exactly one fetch per pull cycle, and 30-15's pending-row UI reads the
// `missing` half of that same snapshot.
import type { WasmUserKey } from "@/lib/crypto";
import { getCollection, type CollectionRow } from "@/lib/vault/api";
import { getFamilyWidePendingSnapshot } from "./familyWidePending";
import { reshareCollectionToNewMember, isConflictError } from "./reseal";
import type { ResealableGrant } from "./api";

/** CR-01 fix (30-REVIEW.md): used only when the collection's OWN
 * `family_wide_access_level` is `null` -- a legacy family-wide collection
 * created before that column existed (a row served before this fix
 * deployed, or mid-rolling-restart). Deliberately never falls back to the
 * caller's OWN held `access_level` -- that used to be exactly this
 * function's bug: a resealer who happens to hold `edit` (e.g. the creator,
 * whose own row is ALWAYS `'edit'` regardless of the level the share was
 * actually created at) would silently upgrade a `read`-declared share to
 * `edit` for every newcomer they resealed to. `"read"` is the safe,
 * never-over-grants default for the genuinely-unknown case. */
const FALLBACK_ACCESS_LEVEL = "read";

/** Per-SESSION set of `"${collectionId}:${recipientUserId}"` pairs this
 * session has already attempted (successfully or not). Bounds repeat work
 * within one session (T-30-21) and makes two overlapping runs in the same
 * tick impossible to double-fire, because a pair is added BEFORE the first
 * await. Cleared on every lock/unlock transition by store.ts via
 * `resetFamilyWideResealAttempts()` below -- mirroring `sync.ts`'s own
 * reset-every-unlock discipline for its latches -- so a pair whose attempt
 * failed transiently is re-attempted on the next unlock's fresh snapshot,
 * never stranded for the lifetime of the tab. */
const attemptedPairs = new Set<string>();

function attemptKey(collectionId: string, recipientUserId: string): string {
  return `${collectionId}:${recipientUserId}`;
}

/** Clears the per-session attempted-pair set. Called from store.ts's
 * `subscribeLockState` unlock branch alongside the other per-unlock latch
 * resets. */
export function resetFamilyWideResealAttempts(): void {
  attemptedPairs.clear();
}

/**
 * For every resealable pair in the current snapshot that this session has
 * not already attempted, reseals the collection's EXISTING Collection Key to
 * that recipient (`reseal.ts`, 30-04) -- no rotation, no server-visible key
 * material, no new endpoint.
 *
 * Never rejects. Each pair's attempt is its own independently-caught
 * `Promise.allSettled` entry, so one transient failure (network drop, a
 * recipient who has not published an identity key yet) can neither block nor
 * abort any other pair's reseal. Partial completion is safe and
 * self-healing: the server's own `INSERT ... ON CONFLICT DO NOTHING` against
 * `collection_keys`'s composite primary key (30-02) makes each grant
 * idempotent, so an interrupted run (a lock mid-batch, a closed tab, two
 * simultaneously-online resealers racing) leaves whatever already landed
 * intact and simply re-attempts the rest from the NEXT unlock's fresh
 * snapshot. That existing idempotency IS the contention story -- this module
 * deliberately invents no coordination scheme of its own.
 */
export async function runFamilyWideResealTrigger(uk: WasmUserKey): Promise<void> {
  const { resealable } = getFamilyWidePendingSnapshot();
  if (resealable.length === 0) {
    // The overwhelmingly common case (no family, or nothing pending):
    // zero extra work on unlock, not even a round trip (T-30-21).
    return;
  }

  // Claim every fresh pair SYNCHRONOUSLY, before the first await below, so a
  // concurrent invocation within the same tick (a WS event and a poll tick
  // landing together) sees them already claimed and fires nothing.
  const fresh: ResealableGrant[] = [];
  for (const grant of resealable) {
    const key = attemptKey(grant.collection_id, grant.recipient_user_id);
    if (attemptedPairs.has(key)) {
      continue;
    }
    attemptedPairs.add(key);
    fresh.push(grant);
  }
  if (fresh.length === 0) {
    return;
  }

  // One `getCollection` per distinct collection per run, shared by every
  // recipient pending on it -- several newcomers on the same family-wide
  // folder cost one round trip, not N.
  const collectionLoads = new Map<string, Promise<CollectionRow>>();
  function loadCollection(collectionId: string): Promise<CollectionRow> {
    let pending = collectionLoads.get(collectionId);
    if (pending === undefined) {
      pending = getCollection(collectionId);
      collectionLoads.set(collectionId, pending);
    }
    return pending;
  }

  await Promise.allSettled(
    fresh.map(async (grant) => {
      try {
        const collection = await loadCollection(grant.collection_id);
        // The current session acts ONLY on keys it already holds. A
        // `sealed_key` of null means this session is on the MISSING side of
        // this collection, not the resealing side -- that case belongs to
        // 30-15's pending-row UI, never to this trigger. (The server's
        // `resealable` query already excludes it; this is the client-side
        // half of the same invariant, cheap because the row is fetched
        // anyway for its access level.)
        if (collection.sealed_key === null || collection.sealed_key === undefined) {
          return;
        }
        // CR-01 fix (30-REVIEW.md): the SHARE's own declared level
        // (`family_wide_access_level`), never `collection.access_level` —
        // the latter is the RESEALER's own `collection_keys` row, which
        // could be `'edit'` (the creator) while the share was declared
        // `'read'`, silently over-granting the newcomer. See
        // `collections::add_member`'s CR-03 fix (`crates/pv-server/src/routes/
        // collections.rs`) for the server-side bound that keeps this call
        // honest even if a caller ever sent something it shouldn't.
        await reshareCollectionToNewMember(
          grant.collection_id,
          grant.recipient_user_id,
          collection.family_wide_access_level ?? FALLBACK_ACCESS_LEVEL,
          uk,
        );
      } catch (err) {
        // CR-03 fix (31-REVIEW.md): `reshareCollectionToNewMember` no longer
        // swallows a 409 itself -- this trigger restores that behavior in
        // its OWN catch, because it is the one caller for whom it is
        // actually correct: this loop's `fresh` set is built from
        // `resealable`, whose own server-side query only ever returns a pair
        // with NO existing `collection_keys` row, so a 409 here can only be
        // a genuine race with another resealer (or a redundant trigger
        // firing twice) landing first -- exactly the state this trigger
        // wants, never a stale-level ambiguity the way a user-chosen level
        // could produce (see `ShareDialog.tsx`'s own 409 handling, which
        // does NOT get this pass).
        if (isConflictError(err)) {
          return;
        }
        // Opportunistic by construction: a failed pair is logged and left
        // for the next unlock's fresh snapshot, never surfaced to the user
        // and never allowed to reach the sync loop.
        console.warn(
          `pv: family-wide lazy reseal failed for collection ${grant.collection_id} -> ${grant.recipient_user_id} -- retrying on a later unlock`,
          err,
        );
      }
    }),
  );
}

File: ios/PasskeyVault/PasskeyVault/Sharing/ResealTrigger.swift
//
//  ResealTrigger.swift
//  PasskeyVault
//
//  Phase 40 (rodzina-i-współdzielenie-na-telefonie), plan 40-10, Task 2.
//  The cadence half of DR-40-B's resealer -- mirrors
//  `web/src/lib/families/resealTrigger.ts` (`git show
//  main:web/src/lib/families/resealTrigger.ts`, header comment included,
//  read this session): runs opportunistically off the existing unlock/sync
//  cycle, is NEVER awaited by anything on the critical path, and tolerates
//  its own failure completely.
//
//  THE TRIGGER SET DELIBERATELY INCLUDES THE SHARER, same as the reference:
//  there is no `recipientUserId != me`-style guard anywhere in this file. A
//  `ResealableGrantRow` names a pair the CALLER already holds a key for and
//  a named member does not -- the server's own `family_wide_pending` query
//  (`families.rs:369-438`) already excludes the caller from ever appearing
//  as a `recipient_user_id` of their own pair, so no such guard is needed
//  here either.
//
//  This file owns exactly TWO things: the per-session attempted-pair set,
//  and the fan-out over a `resealable` snapshot. Fetching that snapshot
//  (`GET /api/families/family-wide-pending`) is deliberately NOT this
//  file's job -- `SharedItemsStore.fetchFamilyWidePending` already owns
//  that call (plan 40-05), and the production wiring (`SyncCoordinator`)
//  passes its `resealable` array straight through, so there is exactly ONE
//  fetch per pull cycle shared by both `PendingKeyState`'s `missing` axis
//  and this trigger's `resealable` fan-out -- mirroring `resealTrigger.ts`'s
//  own "one query, two consumers" header note.
//
//  `actor`, not a plain class: the reference's own load-bearing property is
//  "claim every fresh pair SYNCHRONOUSLY, before the first `await`, so a
//  concurrent invocation within the same tick sees them already claimed".
//  Swift's actor reentrancy model makes this a STRUCTURAL guarantee rather
//  than a discipline someone could get wrong: code between suspension
//  points inside an actor-isolated method runs atomically with respect to
//  every other call into that SAME actor, so the claim loop below (which
//  contains no `await`) can never interleave with another call's claim
//  loop, however the two calls happen to be scheduled. This IS the whole
//  concurrency control (matching the reference's own header: "deliberately
//  invents no coordination scheme of its own") -- no lock, lease, or queue
//  is added anywhere in this file.
//

import Foundation

/// The cadence/fan-out engine over `ResealableGrantRow` (`PendingKeyState.swift`'s
/// own type, decoded from `family_wide_pending`'s `resealable` array).
actor ResealTrigger {
    private var attemptedPairs: Set<String> = []
    private let resealService: ResealService

    init(resealService: ResealService) {
        self.resealService = resealService
    }

    private static func attemptKey(collectionId: String, recipientUserId: String) -> String {
        "\(collectionId):\(recipientUserId)"
    }

    /// Clears the attempted-pair set. Call on EVERY lock AND EVERY unlock
    /// transition (the production wiring's own job, `SyncCoordinator`) --
    /// mirrors `resetFamilyWideResealAttempts()`'s own doc comment: a pair
    /// that failed transiently must be retried on the next unlock's fresh
    /// snapshot, never stranded for the lifetime of this object.
    func resetAttempts() {
        attemptedPairs.removeAll()
    }

    /// One pair's identity, echoed back in `RunOutcome` so a caller (a test,
    /// or a future logging hook) can tell which pair succeeded/failed
    /// without re-deriving the key string this file uses internally.
    struct Attempt: Equatable {
        let collectionId: String
        let recipientUserId: String
    }

    /// One pair whose reseal attempt threw. `error` is NOT `Equatable` --
    /// kept as the real thrown error, never stringified here, so a caller
    /// that cares about the underlying cause (a test asserting on error
    /// type, a log line) can inspect it directly.
    struct FailedAttempt {
        let attempt: Attempt
        let error: Error
    }

    /// The result of one `run` call. NEVER thrown -- `run` has no `throws`
    /// in its own signature, structurally enforcing the reference's "never
    /// rejects" contract: a caller cannot even write a `try` around this
    /// call, let alone have to handle a reseal-specific error type on the
    /// unlock/sync path.
    struct RunOutcome {
        /// The count of FRESH pairs this call actually attempted -- i.e.
        /// `succeeded.count + failed.count`. Zero for an empty `resealable`
        /// list AND for a `resealable` list whose every pair was already
        /// claimed by an earlier call in this session -- the two cases this
        /// file's own header (Pitfall 6, `40-RESEARCH.md`) warns must be
        /// distinguishable from "the trigger ran and delivered nothing":
        /// pairing this field's own zero against a DIFFERENT call's
        /// non-zero value (as `ResealTriggerTests` does) is what makes that
        /// distinction real, since a hard-coded `return` early would make
        /// this field a constant zero forever, not a value that moves.
        let attempted: Int
        let succeeded: [Attempt]
        let failed: [FailedAttempt]
    }

    /// Fans out over `resealable`, claiming each fresh pair SYNCHRONOUSLY
    /// (this file's own header) and then settling each claimed pair
    /// INDEPENDENTLY -- a failure at one pair never prevents any other
    /// pair's attempt, and never causes this function itself to throw. The
    /// server's own `INSERT ... ON CONFLICT DO NOTHING` on `collection_keys`'
    /// composite PK (via `ResealService.reshareCollection`'s own conflict
    /// handling) makes each grant idempotent -- that idempotency IS the
    /// contention story a partial/interrupted run relies on; nothing here
    /// invents a second one.
    @discardableResult
    func run(resealable: [ResealableGrantRow], userKey: FfiUserKey) async -> RunOutcome {
        guard !resealable.isEmpty else {
            // The overwhelmingly common case (no family, or nothing
            // pending): zero extra work, not even a claim-loop iteration.
            return RunOutcome(attempted: 0, succeeded: [], failed: [])
        }

        // Claim every fresh pair SYNCHRONOUSLY, before the first `await`
        // below -- this file's own header explains why this ordering alone
        // is the entire concurrency control.
        var fresh: [ResealableGrantRow] = []
        for grant in resealable {
            let key = Self.attemptKey(collectionId: grant.collection_id, recipientUserId: grant.recipient_user_id)
            guard !attemptedPairs.contains(key) else { continue }
            attemptedPairs.insert(key)
            fresh.append(grant)
        }
        guard !fresh.isEmpty else {
            return RunOutcome(attempted: 0, succeeded: [], failed: [])
        }

        var succeeded: [Attempt] = []
        var failed: [FailedAttempt] = []
        for grant in fresh {
            // CR-03: a lock landing mid-fan-out (the caller cancels the
            // task that owns this call) stops the REMAINING pairs instead
            // of draining the whole list -- pairs already claimed in
            // `attemptedPairs` above stay claimed, so they are retried
            // cleanly on the next unlock's `resetAttempts()` + fresh
            // snapshot rather than either double-attempted or silently
            // dropped forever.
            guard !Task.isCancelled else { break }
            let attempt = Attempt(collectionId: grant.collection_id, recipientUserId: grant.recipient_user_id)
            do {
                try await resealService.reshareCollection(
                    collectionId: grant.collection_id, recipientUserId: grant.recipient_user_id, userKey: userKey
                )
                succeeded.append(attempt)
            } catch {
                // Opportunistic by construction: a failed pair is recorded
                // here and left for the next unlock's fresh snapshot (once
                // `resetAttempts()` runs), never surfaced to the user and
                // never allowed to abort the rest of this fan-out.
                failed.append(FailedAttempt(attempt: attempt, error: error))
            }
        }
        return RunOutcome(attempted: fresh.count, succeeded: succeeded, failed: failed)
    }
}
